### ISSUE TITLE
Warnings during build-python

### ISSUE THREAD

**Issue by brockho** [numbbo/coco#1350]
We still have some warnings that show up when building the python wrapper of COCO. I guess that the same appear when compiling C with a strict enough compiler:

```
code-experiments/src/suite_bbob_legacy_code.c(84) : warning C4018: '<' : signed/unsigned mismatch
code-experiments/src/suite_bbob_legacy_code.c(145) : warning C4018: '<' : signed/unsigned mismatch
code-experiments/src/suite_bbob_legacy_code.c(148) : warning C4018: '<' : signed/unsigned mismatch
code-experiments/src/suite_bbob_legacy_code.c(150) : warning C4018: '<' : signed/unsigned mismatch
code-experiments/src/suite_bbob_legacy_code.c(154) : warning C4018: '<' : signed/unsigned mismatch
code-experiments/src/suite_bbob_legacy_code.c(156) : warning C4018: '<' : signed/unsigned mismatch
code-experiments/src/suite_bbob_legacy_code.c(180) : warning C4018: '<' : signed/unsigned mismatch
code-experiments/src/large_scale_transformations.c(181) : warning C4018: '<' : signed/unsigned misma
tch
code-experiments/src/large_scale_transformations.c(215) : warning C4018: '<' : signed/unsigned misma
tch
code-experiments/src/large_scale_transformations.c(225) : warning C4018: '<' : signed/unsigned misma
tch
code-experiments/src/large_scale_transformations.c(291) : warning C4018: '<' : signed/unsigned misma
tch
code-experiments/src/transform_vars_permblockdiag.c(84) : warning C4018: '<' : signed/unsigned misma
tch
code-experiments/src/transform_vars_permblockdiag.c(99) : warning C4018: '<' : signed/unsigned misma
tch
code-experiments/src/transform_vars_permblockdiag.c(100) : warning C4018: '>=' : signed/unsigned mis
match
code-experiments/src/f_sharp_ridge.c(32) : warning C4244: '=' : conversion from 'double' to 'size_t'
, possible loss of data
code-experiments/src/c_linear.c(229) : warning C4244: '=' : conversion from 'double' to 'size_t', po
ssible loss of data
code-experiments/src/suite_cons_bbob.c(64) : warning C4090: 'function' : different 'const' qualifier
s
code-experiments/src/suite_cons_bbob.c(130) : warning C4090: 'function' : different 'const' qualifie
rs
code-experiments/src/suite_cons_bbob.c(135) : warning C4090: 'function' : different 'const' qualifie
rs
code-experiments/src/coco_suite.c(495) : warning C4018: '<' : signed/unsigned mismatch
code-experiments/src/coco_suite.c(508) : warning C4018: '<' : signed/unsigned mismatch
code-experiments/src/logger_bbob.c(615) : warning C4090: 'function' : different 'const' qualifiers
code-experiments/src/logger_bbob.c(615) : warning C4028: formal parameter 2 different from declarati
on
[...]
```

I actually got to see the warnings due to another error (`phrsampaio-master-testmerge` branch on Windows). It will be nice to get rid of most of the warnings at some point in time.

**Comment by nikohansen** [comment]
I consider this critical, because superfluous warnings tend to obscure important warnings and are hence a potential source of large time sinks.

**Comment by nikohansen** [comment]
Please everybody consider addressing those warnings coming from your own code.

**Comment by nikohansen** [comment]
The warnings in file `large_scale_transformations.c` should be gone with one of the next commits to the development branch.

**Comment by brockho** [comment]
The latest pull request should have removed all but one warnings (at least only one shows up in my `gcc`). The last one, we shall discuss tomorrow in our weekly BBOB/COCO meeting.

**Comment by nikohansen** [comment]
There still remain four, two of which should still be addressed. FTR, and how to get these warnings displayed:

```
[12:00:28] 0 coco$ ./do.py run-c  # build (amalgamate) the current code
AML	['code-experiments/src/coco_random.c', 'code-experiments/src/coco_suite.c', 'code-experiments/src/coco_observer.c', 'code-experiments/src/coco_archive.c', 'code-experiments/src/coco_runtime_c.c'] -> code-experiments/build/c/coco.c
EXPAND	code-experiments/build/c/coco.c.in to code-experiments/build/c/coco.c
EXPAND	code-experiments/src/coco.h to code-experiments/build/c/coco.h
COPY	code-experiments/build/c/coco.c -> code-experiments/examples/bbob2009-c-cmaes/coco.c
EXPAND	code-experiments/build/c/coco.h to code-experiments/examples/bbob2009-c-cmaes/coco.h
WRITE	code-experiments/build/c/REVISION
WRITE	code-experiments/build/c/VERSION
MAKE	clean in code-experiments/build/c
COPY	Makefile.in -> Makefile
MAKE	all in code-experiments/build/c
COPY	Makefile.in -> Makefile
RUN	./example_experiment in code-experiments/build/c
[12:00:58] 0 coco$ cd code-experiments/build/c/
[12:01:15] 0 c$ make clean
rm -f coco.o 
rm -f example_experiment.o example_experiment 
[12:01:18] 0 c$ make
cc -c -g -ggdb -std=c89 -pedantic -Wall -Wextra -Wstrict-prototypes -Wshadow -Wno-sign-compare -Wconversion -o example_experiment.o example_experiment.c
example_experiment.c:184:12: warning: ISO C90 forbids mixing declarations and code [-Wdeclaration-after-statement]
      long evaluations_remaining = (long) (dimension * BUDGET_MULTIPLIER) - (long) evaluations_done;
           ^
1 warning generated.
cc -c -g -ggdb -std=c89 -pedantic -Wall -Wextra -Wstrict-prototypes -Wshadow -Wno-sign-compare -Wconversion -o coco.o coco.c
code-experiments/src/coco_problem.c:463:16: warning: comparison of array 'problem->final_target_delta' equal to a null pointer is always false
      [-Wtautological-pointer-compare]
      problem->final_target_delta == NULL ||
      ~~~~~~~~~^~~~~~~~~~~~~~~~~~    ~~~~
code-experiments/src/coco_problem.c:464:16: warning: comparison of array 'problem->best_observed_fvalue' equal to a null pointer is always false
      [-Wtautological-pointer-compare]
      problem->best_observed_fvalue == NULL)
      ~~~~~~~~~^~~~~~~~~~~~~~~~~~~~    ~~~~
code-experiments/src/logger_bbob.c:627:30: warning: incompatible pointer types assigning to 'coco_evaluate_function_t' (aka 'void (*)(struct coco_problem_s *,
      const double *, double *)') from 'void (coco_problem_t *, double *, double *)' (aka 'void (struct coco_problem_s *, double *, double *)')
      [-Wincompatible-pointer-types]
  problem->evaluate_function = logger_bbob_evaluate;
                             ^ ~~~~~~~~~~~~~~~~~~~~
3 warnings generated.
cc -g -ggdb -std=c89 -pedantic -Wall -Wextra -Wstrict-prototypes -Wshadow -Wno-sign-compare -Wconversion -o example_experiment coco.o example_experiment.o -lm  
[12:01:20] 0 c$ 
```

**Comment by nikohansen** [comment]
Down to 3 warnings:
```
[11:41:54] 0 c$ make
cc -c -g -ggdb -std=c89 -pedantic -Wall -Wextra -Wstrict-prototypes -Wshadow -Wno-sign-compare -Wconversion -o example_experiment.o example_experiment.c
cc -c -g -ggdb -std=c89 -pedantic -Wall -Wextra -Wstrict-prototypes -Wshadow -Wno-sign-compare -Wconversion -o coco.o coco.c
code-experiments/src/coco_problem.c:463:16: warning: comparison of array 'problem->final_target_delta' equal to a null pointer is always false
      [-Wtautological-pointer-compare]
      problem->final_target_delta == NULL ||
      ~~~~~~~~~^~~~~~~~~~~~~~~~~~    ~~~~
code-experiments/src/coco_problem.c:464:16: warning: comparison of array 'problem->best_observed_fvalue' equal to a null pointer is always false
      [-Wtautological-pointer-compare]
      problem->best_observed_fvalue == NULL)
      ~~~~~~~~~^~~~~~~~~~~~~~~~~~~~    ~~~~
code-experiments/src/logger_bbob.c:627:30: warning: incompatible pointer types assigning to 'coco_evaluate_function_t' (aka 'void (*)(struct coco_problem_s *,
      const double *, double *)') from 'void (coco_problem_t *, double *, double *)' (aka 'void (struct coco_problem_s *, double *, double *)')
      [-Wincompatible-pointer-types]
  problem->evaluate_function = logger_bbob_evaluate;
                             ^ ~~~~~~~~~~~~~~~~~~~~
3 warnings generated.
```

**Comment by nikohansen** [comment]
I hesitate to remove the last two warnings by removing the (currently useless) code: if someone changed `problem->final_target_delta` from a `double[1]` to a `*double`, the code would make sense and be rather useful. I am OK how it is now, despite the two remaining warnings.

**Comment by brockho** [comment]
We discussed today that we can actually close this issue. It is not an actual problem because the tested equality can never happen. However, if we decide later on to change the code from `double[1]` to `*double` we would like to have exactly this code, hence, we leave it in.

**Comment by ttusar** [comment]
After already closing the issue, @brockho and @nikohansen kindly conceded to my wish to clear all warnings. This was fixed with commit #1541 (a comment was added near variable declaration to remind that a change from array to pointer requires NULL checks).
